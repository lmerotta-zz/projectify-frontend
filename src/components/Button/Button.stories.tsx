// Button.stories.tsx
import { ComponentProps } from "react";
import Button from "./Button";
import { Story } from "@storybook/react/types-6-0";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Button",
  component: Button,
  argTypes: {
    disabled: { control: { type: "boolean" } },
  },
};

type Props = Omit<ComponentProps<typeof Button>, "as">;
const Template: Story<Props> = (args) => <Button {...args}>My Button</Button>;

export const Default = Template.bind({});
