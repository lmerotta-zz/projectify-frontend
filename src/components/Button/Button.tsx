/** @jsxImportSource @emotion/react */
import { ComponentProps } from "react";
import tw, { styled } from "twin.macro";

type ButtonColor = "primary" | "secondary";

type ButtonProps = ComponentProps<"button"> & {
  color?: ButtonColor;
  block?: boolean;
};

const Button = styled.button<ButtonProps>`
  ${tw`px-3 py-3 text-white transition duration-150`}
  ${({ color = "primary" }) =>
    color === "primary"
      ? tw`bg-primary hover:bg-primary-hover`
      : tw`bg-secondary hover:bg-secondary-hover`}

  ${({ block }) => block && tw`w-full`}

  ${({ disabled }) => disabled && tw`text-gray-300`}
  ${({ disabled, color = "primary" }) => {
    if (disabled) {
      switch (color) {
        case "primary":
          return tw`bg-primary-dark hover:bg-primary-dark`;
        case "secondary":
          return tw`bg-secondary-dark hover:bg-secondary-dark`;
      }
    }
  }}
`;

export default Button;
