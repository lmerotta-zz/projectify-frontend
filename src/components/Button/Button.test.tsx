import { render } from "@testing-library/react";
import Button from "./Button";
import each from "jest-each";

describe("Button unit tests", () => {
  each([["primary"], ["secondary"]]).it("Renders with color %s", (color) => {
    const { asFragment } = render(
      <>
        <Button color={color}>My Button</Button>
        <Button color={color} disabled>
          My disabled Button
        </Button>
      </>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
