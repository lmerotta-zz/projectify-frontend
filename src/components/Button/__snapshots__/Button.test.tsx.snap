// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button unit tests Renders with color primary 1`] = `
<DocumentFragment>
  .emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(147, 51, 234, var(--tw-bg-opacity));
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-0:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(126, 34, 206, var(--tw-bg-opacity));
}

<button
    class="emotion-0"
    color="primary"
  >
    My Button
  </button>
  .emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(147, 51, 234, var(--tw-bg-opacity));
  --tw-text-opacity: 1;
  color: rgba(209, 213, 219, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(88, 28, 135, var(--tw-bg-opacity));
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-0:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(126, 34, 206, var(--tw-bg-opacity));
}

.emotion-0:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(88, 28, 135, var(--tw-bg-opacity));
}

<button
    class="emotion-0"
    color="primary"
    disabled=""
  >
    My disabled Button
  </button>
</DocumentFragment>
`;

exports[`Button unit tests Renders with color secondary 1`] = `
<DocumentFragment>
  .emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(234, 179, 8, var(--tw-bg-opacity));
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-0:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(202, 138, 4, var(--tw-bg-opacity));
}

<button
    class="emotion-0"
    color="secondary"
  >
    My Button
  </button>
  .emotion-0 {
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(234, 179, 8, var(--tw-bg-opacity));
  --tw-text-opacity: 1;
  color: rgba(209, 213, 219, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(133, 77, 14, var(--tw-bg-opacity));
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-0:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(202, 138, 4, var(--tw-bg-opacity));
}

.emotion-0:hover {
  --tw-bg-opacity: 1;
  background-color: rgba(133, 77, 14, var(--tw-bg-opacity));
}

<button
    class="emotion-0"
    color="secondary"
    disabled=""
  >
    My disabled Button
  </button>
</DocumentFragment>
`;
